Number_types/include/CGAL/Quotient.h

```diff
@@ -686,11 +707,234 @@ template < class NT > class Real_embeddable_traits_quotient_base< Quotient<NT> >
+            x.num = -x.num;
+            x.den = -x.den;
+          } else if (x.num < 0 && x.den > 0) {
+            change_sign = true;
+            x.num = -x.num;
+          } else if (x.num > 0 && x.den < 0) {
+            change_sign = true;
+            x.den = -x.den;
+          }
+          CGAL_assertion(x.num > 0 && x.den > 0);
+
+          const auto num = get_interval_exp(x.num);
+          const auto den = get_interval_exp(x.den);
+
+          const Interval_nt<> div = num.first / den.first;
+          const int64_t e = num.second - den.second;
+          std::tie(l, u) = ldexp(div, e).pair();
+
+          if (change_sign) {
+            const double t = l;
+            l = -u;
+            u = -t;
+          }
+
+          CGAL_assertion(are_correct_bounds(l, u, input));
+          return std::make_pair(l, u);
+        }
+
+        std::pair<double, double> get_interval_as_boost( Type x ) const {
+
+          CGAL_assertion_code(const Type input = x);
+          double l = 0.0, u = 0.0;
+          if (x.num == 0) { // return [0.0, 0.0]
+            CGAL_assertion(are_correct_bounds(l, u, input));
+            return std::make_pair(l, u);
+          }
+          CGAL_assertion(x.num != 0);
+          CGAL_assertion(x.den != 0);
+
+          // Handle signs.
+          bool change_sign = false;
+          if (x.num < 0 && x.den < 0) {
+            x.num = -x.num;
+            x.den = -x.den;
+          } else if (x.num < 0 && x.den > 0) {
+            change_sign = true;
+            x.num = -x.num;
+          } else if (x.num > 0 && x.den < 0) {
+            change_sign = true;
+            x.den = -x.den;
+          }
+          CGAL_assertion(x.num > 0 && x.den > 0);
+
+          const int64_t num_dbl_digits = std::numeric_limits<double>::digits - 1;
+          const int64_t msb_num = static_cast<int64_t>(msb(x.num));
+          const int64_t msb_den = static_cast<int64_t>(msb(x.den));
+          const int64_t msb_diff = msb_num - msb_den;
+          const int64_t shift = num_dbl_digits - msb_diff;
+
+          if (shift > 0) {
+            CGAL_assertion(msb_diff < num_dbl_digits);
+            x.num <<= shift;
+          } else if (shift < 0) {
+            CGAL_assertion(msb_diff > num_dbl_digits);
+            x.den <<= boost::multiprecision::detail::unsigned_abs(shift);
+          }
+          CGAL_assertion(num_dbl_digits ==
+            static_cast<int64_t>(msb(x.num)) - static_cast<int64_t>(msb(x.den)));
+
+          decltype(x.num) p, q, r;
+          boost::multiprecision::divide_qr(x.num, x.den, q, r);
+          const int64_t q_bits = static_cast<int64_t>(msb(q));
+
+          CGAL_assertion(q_bits == num_dbl_digits || r != 0);
+          if (r != 0) {
+            p = q;
+            ++q;
+          } else {
+            p = q;
+          }
+
+          CGAL_assertion(p >= 0 && q >= 0);
+          const uint64_t pp = static_cast<uint64_t>(p);
+          const uint64_t qq = static_cast<uint64_t>(q);
```

Comment:
In case this is expensive we can avoid doing it twice in case `p == q`.

Reply:
True. Fixed.